### PR TITLE
rename `broad-except` and new check `broad-exception-raised`

### LINKED ISSUE #7494: Check for raise Exception or BaseException
### Current problem

Linting a codebase with a lot of `raise Exception()` statements. This is a bad practice as it makes it impossible to write narrow exception clauses.

### Desired solution

Forgive me if I missed it, but as far as I can tell, a check doesn't currently exist. The check should fail when an instance of `Exception` or `BaseException` is raised directly.

### PR DESCRIPTION
**Pull request by clavedeluna**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

This PR renames `broad-except` to `broad-exception-caught`:

```
    except Exception:  # [broad-except]
        pass
```
becomes

```
    except Exception:  # [broad-exception-caught]
        pass
```

and adds a new checker, `broad-exception-raised`

Closes #7494

### CODE REVIEW

File: pylint/checkers/exceptions.py [hunk 1]
```diff
@@ -555,7 +566,9 @@ def visit_tryexcept(self, node: nodes.TryExcept) -> None:
                         and not _is_raising(handler.body)
                     ):
                         self.add_message(
-                            "broad-except", args=exception.name, node=handler.type
+                            "broad-exception-caught",
+                            args=exception.name,
+                            node=handler.type,
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/exceptions.py:571]:
```suggestion
                            node=handler.type,
                            confidence=interface.HIGH,
```
AFAIK.

**Review by Pierre-Sassoulas — COMMENTED** [review]
This looks pretty good and polished at first glance 👌

File: pylint/checkers/exceptions.py (hunk 1 above)
Comment (clavedeluna) [reply pylint/checkers/exceptions.py:571]: Correct. Have this in a commit not pushed. I made a draft PR to see how the primer reacted. Not ready for review yet.

**Comment by clavedeluna** [comment]
> 🤖 **Effect of this PR on checked open source code:** 🤖
> 
> **Effect on [astroid](https://github.com/PyCQA/astroid):** The following messages are now emitted:
> 
> The following messages are no longer emitted:
> 
> **Effect on [black](https://github.com/psf/black):** The following messages are now emitted:
> 
> The following messages are no longer emitted:
> 
> **Effect on [django](https://github.com/django/django):** The following messages are now emitted:
> 
> The following messages are no longer emitted:
> 
> **Effect on [flask](https://github.com/pallets/flask):** The following messages are now emitted:
> 
> The following messages are no longer emitted:
> 
> **Effect on [pandas](https://github.com/pandas-dev/pandas):** The following messages are now emitted:
> 
> The following messages are no longer emitted:
> 
> **Effect on [psycopg](https://github.com/psycopg/psycopg):** The following messages are now emitted:
> 
> The following messages are no longer emitted:
> 
> **Effect on [pygame](https://github.com/pygame/pygame):** The following messages are now emitted:
> 
> The following messages are no longer emitted:
> 
> **Effect on [pytest](https://github.com/pytest-dev/pytest):** The following messages are now emitted:

results look reasonable, my question about the `broad-except` deprecation, is will repos that do this
```
    except Exception:  # pylint: disable=broad-except
```
or other forms of disabling `broad-except`, now need to change the disable to `disable=broad-exception-caught`? or will `old_names` handle this?

**Comment by Pierre-Sassoulas** [comment]
> or will old_names handle this?

That's the use case for old_names :)

File: pylint/checkers/exceptions.py [hunk 2]
```diff
@@ -225,14 +245,28 @@ def visit_instance(self, instance: objects.ExceptionInstance) -> None:
...
+                    confidence=INFERENCE,
+                )
 
     def visit_tuple(self, _: nodes.Tuple) -> None:
-        self._checker.add_message("raising-bad-type", node=self._node, args="tuple")
+        self._checker.add_message(
+            "raising-bad-type",
+            node=self._node,
+            args="tuple",
+            confidence=INFERENCE,
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/exceptions.py:259]: Why inference and not high? Also it should be done in another MR.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Look good apart for the unrelated confidence change :) We also probably need to regenerate the doc following the removal of a message (``tox -e docs`` or ``cd doc;make build``)

File: pylint/checkers/exceptions.py (hunk 2 above)
Comment (clavedeluna) [reply pylint/checkers/exceptions.py:259]: It's in a sep commit but np, I can split that out. Also, I'm pretty sure I traced all the code / funcs that received inference results, but I could of course have made a mistake. Can move it to a sep PR for discussion.

Comment (Pierre-Sassoulas) [reply pylint/checkers/exceptions.py:259]: Thank you !

**Review by Pierre-Sassoulas — COMMENTED** [review]
I realized there were too much issue with the doc right now (because of https://github.com/PyCQA/pylint/pull/6736 being blocked) and created a MR to fix it (#7711). I'm going to review again when the other branch is merged and this one is rebased on it.

File: doc/data/messages/b/broad-exception-raised/good.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
+def small_apple(apple, length):
+    if len(apple) < length:
+        raise IndexError("Apple is too small!")
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/b/broad-exception-raised/good.py:3]:
```suggestion
        raise ValueError("Apple is too small!")
```
Semantically I think it make more sense.

File: doc/data/messages/m/magic-value-comparison/details.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+You can help us make the doc better `by contributing <https://github.com/PyCQA/pylint/issues/5953>`_ !
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/m/magic-value-comparison/details.rst:1]: I think there's still a problem here following resolution of conflict for the merge of main I just did. This new file should not exists. Would you mind rebasing on main and cleaning up the doc generation commit by doing a partial add of only the broad-except-x related changes ? Sorry for the mess, we did not have time to fix #6736 yet or this would not be a problem.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

File: doc/data/messages/m/magic-value-comparison/details.rst (hunk 4 above)
Comment (clavedeluna) [reply doc/data/messages/m/magic-value-comparison/details.rst]: done, I think the latest looks most accurate, let's see how CI reacts

**Review by Pierre-Sassoulas — APPROVED** [review]
Pretty clean check and refactor, thank you 👌

**Review by DudeNr33 — COMMENTED** [review]
Only looked at the `pyreverse` change, but that looks good to me!

**Comment by cdce8p** [comment]
Just a suggestion for the future. Please split it into two separate PRs. I would consider a new check a larger enough change that the separation would probably be better, even if it's related to the refactoring.
